### ISSUE TITLE
Cannot iterate over an index set constructed with `Maybe`

### ISSUE THREAD

**Issue by axch** [google-research/dex-lang#530]
```
for i:(Maybe (Fin 2)). 1
Compiler bug!
Please report this at github.com/google-research/dex-lang/issues

Not implemented (Maybe (Fin 2))
CallStack (from HasCallStack):
  error, called at src/lib/Builder.hs:860:20 in dex-0.1.0.0-3TQe6uju44BJFUXpaseRHY:Builder
```

(in my checkout, that source location is the default case of `Builder.indexSetSizeE`).

And there I was thinking I could be clever about how to draw from a deck that contains a special "reshuffle the deck and draw again" card that behaves differently from all other cards...

**Comment by apaszke** [comment]
Yeah this is a know issue at the moment (will be resolved once we get to user-defined index sets), but for the time being you can get the same trick to work using variant types. Just use `{a: Unit | b: (Fin 2)}`. You can use more descriptive names of course, but watch out because the order of the index set is based on a lexicographic sort of the field names. Because of this unclear ordering between fields, we're thinking of phasing out variant types as index sets, but it's the only sum type that works for now.
